diff --git a/QUDAConfig.cmake.in b/QUDAConfig.cmake.in
index 289c235..ba699d7 100644
--- a/QUDAConfig.cmake.in
+++ b/QUDAConfig.cmake.in
@@ -6,6 +6,8 @@ set(QUDA_QMP @QUDA_QMP@)
 set(QUDA_MPI @QUDA_MPI@)
 set(QUDA_QIO @QUDA_QIO@)
 set(QUDA_OPENMP @QUDA_OPENMP@)
+set(QUDA_VERSION @CMAKE_PROJECT_VERSION@)
+set(GITVERSION @GITVERSION@)
 
 if( QUDA_QMP AND QUDA_MPI )
   message(FATAL_ERROR "Cannot have both QMP and MPI configured")  
