diff --git a/lib/copy_gauge.cpp b/lib/copy_gauge.cpp
index 07f0f9f..cc976be 100644
--- a/lib/copy_gauge.cpp
+++ b/lib/copy_gauge.cpp
@@ -21,7 +21,7 @@ namespace quda {
 
   void checkMomOrder(const GaugeField &u) {
     if (u.Order() == QUDA_FLOAT2_GAUGE_ORDER) {
-      if (u.Reconstruct() != QUDA_RECONSTRUCT_10)
+      if (u.Reconstruct() != QUDA_RECONSTRUCT_10 and u.Reconstruct() != QUDA_RECONSTRUCT_NO)
 	errorQuda("Unsuported order %d and reconstruct %d combination", u.Order(), u.Reconstruct());
     } else if (u.Order() == QUDA_TIFR_GAUGE_ORDER || u.Order() == QUDA_TIFR_PADDED_GAUGE_ORDER) {
       if (u.Reconstruct() != QUDA_RECONSTRUCT_NO)
diff --git a/lib/copy_gauge_inc.cu b/lib/copy_gauge_inc.cu
index 223ba2d..b644593 100644
--- a/lib/copy_gauge_inc.cu
+++ b/lib/copy_gauge_inc.cu
@@ -300,10 +300,18 @@ namespace quda {
         // momentum only currently supported on MILC (10), TIFR (18) and Float2 (10) fields currently
 	if (out.Order() == QUDA_FLOAT2_GAUGE_ORDER) {
 	  if (in.Order() == QUDA_FLOAT2_GAUGE_ORDER) {
-	    typedef FloatNOrder<FloatOut,10,2,10> momOut;
-	    typedef FloatNOrder<FloatIn,10,2,10> momIn;
-	    CopyGaugeArg<FloatOut, FloatIn, 10, momOut, momIn> arg(momOut(out, Out, 0), momIn(in, In, 0), in);
-	    copyMom<decltype(arg)>(arg,out,in,location);
+	    if (out.Reconstruct() == QUDA_RECONSTRUCT_10) {
+	      typedef FloatNOrder<FloatOut,10,2,10> momOut;
+	      typedef FloatNOrder<FloatIn,10,2,10> momIn;
+	      CopyGaugeArg<FloatOut, FloatIn, 10, momOut, momIn> arg(momOut(out, Out, 0), momIn(in, In, 0), in);
+	      copyMom<decltype(arg)>(arg,out,in,location);
+	    } else if (out.Reconstruct() == QUDA_RECONSTRUCT_NO) {
+	      typedef FloatNOrder<FloatOut,18,2,18> momOut;
+	      typedef FloatNOrder<FloatIn,18,2,11> momIn;
+	      CopyGaugeArg<FloatOut, FloatIn, 18, momOut, momIn> arg(momOut(out, Out, 0), momIn(in, In, 0), in);
+	      copyMom<decltype(arg)>(arg,out,in,location);
+	    }
+	    
 	  } else if (in.Order() == QUDA_QDP_GAUGE_ORDER) {
 #ifdef BUILD_QDP_INTERFACE
 	    typedef FloatNOrder<FloatOut,10,2,10> momOut;
